### PR TITLE
TTY: serialize tty_write() calls and implement draining the output queue.

### PR DESCRIPTION
**Pull request by j-piecuch**

### CODE REVIEW

File: sys/kern/tty.c [hunk 1]
```diff
@@ -555,6 +569,38 @@ static void tty_output(tty_t *tty, uint8_t c) {
...
+
+/*
+ * Write a single character to the terminal.
+ * If it can't immediately write the character due to lack of space
+ * in the output queue, it goes to sleep waiting for space to become available.
+ */
+static void tty_output_sleep(tty_t *tty, uint8_t c) {
+  while (!tty_output(tty, c)) {
+    tty_notify_out(tty);
+    if (tty->t_outq.count < TTY_OUT_LOW_WATER)
```
Comment (cahirwpz) [inline sys/kern/tty.c:583]: What's the purpose of this line. At first glance I'd say this condition causes busy-looping...

File: sys/kern/tty.c [hunk 2]
```diff
@@ -563,21 +609,47 @@ static int tty_write(vnode_t *v, uio_t *uio, int ioflags) {
...
-      tty->t_rocount = 0;
-      tty_output(tty, c);
-    }
-    tty_notify_out(tty);
+    /* Wait for our turn. */
+    while (tty->t_flags & TF_OUT_BUSY)
+      cv_wait(&tty->t_serialize_cv, &tty->t_lock);
+    tty->t_flags |= TF_OUT_BUSY;
+
+    error = tty_do_write(tty, uio);
```
Comment (cahirwpz) [inline sys/kern/tty.c:618]: Perhaps a comment would be helpful here. It's not obvious that `tty_do_write` may release `t_lock` and setting `TF_OUT_BUSY` is well justified.

File: sys/kern/tty.c [hunk 3]
```diff
@@ -563,21 +609,45 @@ static int tty_write(vnode_t *v, uio_t *uio, int ioflags) {
...
+
+    tty->t_flags &= ~TF_OUT_BUSY;
+    cv_signal(&tty->t_serialize_cv);
   }
 
   return error;
 }
 
+void tty_getc_done(tty_t *tty) {
```
Comment (cahirwpz) [inline sys/kern/tty.c:627]: I simplified condition checking in this procedure. Please check if I did not mess up.

**Review by cahirwpz — COMMENTED** [review]
I'm somewhat concerned about `tty_output_sleep`. Otherwise looks mostly ok.

**Comment by cahirwpz** [comment]
Please resolve conflicts.

File: sys/kern/tty.c (hunk 1 above)
Comment (j-piecuch) [reply sys/kern/tty.c:583]: `tty_notify_out` may output characters synchronously. This line checks if it has output enough characters for us not to need to wait.

File: sys/kern/tty.c (hunk 3 above)
Comment (j-piecuch) [reply sys/kern/tty.c:627]: LGTM

File: sys/kern/tty.c (hunk 1 above)
Comment (cahirwpz) [reply sys/kern/tty.c:584]: Such comment here would be helpful while reading.

**Comment by cahirwpz** [comment]
If comments are added, there'll be nothing else to stop use from merging the PR with `master`.

File: sys/kern/tty.c (hunk 2 above)
Comment (j-piecuch) [reply sys/kern/tty.c:623]: Done.

File: sys/kern/tty.c (hunk 1 above)
Comment (j-piecuch) [reply sys/kern/tty.c:586]: Done.

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**